File: data/snippets/quiz012.cpp
// 当需要添加新的会员等级时，需要修改多个类和方法

class DiscountCalculator {
public:
    double calculateDiscount(double amount, std::string memberLevel) {
        if (memberLevel == "bronze") {
            return amount * 0.05; // 铜牌会员 5% 折扣
        } else if (memberLevel == "silver") {
            return amount * 0.10; // 银牌会员 10% 折扣
        } else if (memberLevel == "gold") {
            return amount * 0.15; // 金牌会员 15% 折扣
        }
        return 0;
    }
};

class RewardPointsCalculator {
public:
    int calculatePoints(double purchaseAmount, std::string memberLevel) {
        int basePoints = static_cast<int>(purchaseAmount);
        
        if (memberLevel == "bronze") {
            return basePoints * 1; // 铜牌会员 1倍积分
        } else if (memberLevel == "silver") {
            return basePoints * 2; // 银牌会员 2倍积分
        } else if (memberLevel == "gold") {
            return basePoints * 3; // 金牌会员 3倍积分
        }
        return basePoints;
    }
};

class MemberBenefitDisplay {
public:
    std::string getBenefitDescription(std::string memberLevel) {
        if (memberLevel == "bronze") {
            return "铜牌会员福利: 5%折扣, 1倍积分, 普通客服";
        } else if (memberLevel == "silver") {
            return "银牌会员福利: 10%折扣, 2倍积分, 优先客服";
        } else if (memberLevel == "gold") {
            return "金牌会员福利: 15%折扣, 3倍积分, VIP客服, 生日礼物";
        }
        return "非会员无特殊福利";
    }
};

class CustomerService {
public:
    int getResponsePriority(std::string memberLevel) {
        if (memberLevel == "bronze") {
            return 3; // 普通优先级
        } else if (memberLevel == "silver") {
            return 2; // 中等优先级
        } else if (memberLevel == "gold") {
            return 1; // 最高优先级
        }
        return 4; // 非会员优先级最低
    }
    
    bool hasVIPHotline(std::string memberLevel) {
        return memberLevel == "gold"; // 只有金牌会员有VIP热线
    }
};

// 其他可能需要修改的类...

// 客户端使用示例
class CheckoutProcess {
public:
    void processCheckout(double amount, std::string memberLevel) {
        DiscountCalculator discountCalc;
        RewardPointsCalculator pointsCalc;
        MemberBenefitDisplay benefitDisplay;
        
        double discount = discountCalc.calculateDiscount(amount, memberLevel);
        int points = pointsCalc.calculatePoints(amount, memberLevel);
        
        std::cout << "订单金额: " << amount << std::endl;
        std::cout << "折扣: " << discount << std::endl;
        std::cout << "应付金额: " << (amount - discount) << std::endl;
        std::cout << "获得积分: " << points << std::endl;
        std::cout << benefitDisplay.getBenefitDescription(memberLevel) << std::endl;
    }
};

File: data/snippets/quiz012.java
/// Imagine when adding a new membership level, what changes would be needed?

class DiscountCalculator {
    public double calculateDiscount(double amount, String memberLevel) {
        if ("bronze".equals(memberLevel)) {
            return amount * 0.05; // Bronze member 5% discount
        } else if ("silver".equals(memberLevel)) {
            return amount * 0.10; // Silver member 10% discount
        } else if ("gold".equals(memberLevel)) {
            return amount * 0.15; // Gold member 15% discount
        } else if ("platinum".equals(memberLevel)) {
            return amount * 0.20; // Platinum member 20% discount
        }
        return 0;
    }
}

class RewardPointsCalculator {
    public int calculatePoints(double purchaseAmount, String memberLevel) {
        int basePoints = (int) purchaseAmount;
        
        if ("bronze".equals(memberLevel)) {
            return basePoints * 1; // Bronze member 1x points
        } else if ("silver".equals(memberLevel)) {
            return basePoints * 2; // Silver member 2x points
        } else if ("gold".equals(memberLevel)) {
            return basePoints * 3; // Gold member 3x points
        } else if ("platinum".equals(memberLevel)) {
            return basePoints * 4; // Platinum member 4x points
        }
        return basePoints;
    }
}

class MemberBenefitDisplay {
    public String getBenefitDescription(String memberLevel) {
        if ("bronze".equals(memberLevel)) {
            return "Bronze member benefits: 5% discount, 1x points, standard support";
        } else if ("silver".equals(memberLevel)) {
            return "Silver member benefits: 10% discount, 2x points, priority support";
        } else if ("gold".equals(memberLevel)) {
            return "Gold member benefits: 15% discount, 3x points, VIP support, birthday gift";
        } else if ("platinum".equals(memberLevel)) {
            return "Platinum member benefits: 20% discount, 4x points, dedicated support, exclusive events";
        }
        return "Non-member: no special benefits";
    }
}

class CustomerService {
    public int getResponsePriority(String memberLevel) {
        if ("bronze".equals(memberLevel)) {
            return 3; // Standard priority
        } else if ("silver".equals(memberLevel)) {
            return 2; // Medium priority
        } else if ("gold".equals(memberLevel)) {
            return 1; // High priority
        } else if ("platinum".equals(memberLevel)) {
            return 0; // Highest priority
        }
        return 4; // Non-member lowest priority
    }
    
    public boolean hasVIPHotline(String memberLevel) {
        return "gold".equals(memberLevel) || "platinum".equals(memberLevel);
    }
    
    public boolean hasDedicatedManager(String memberLevel) {
        return "platinum".equals(memberLevel);
    }
}

class ShippingService {
    public double getShippingDiscount(String memberLevel) {
        if ("bronze".equals(memberLevel)) {
            return 0.0; // No shipping discount
        } else if ("silver".equals(memberLevel)) {
            return 0.5; // 50% shipping discount
        } else if ("gold".equals(memberLevel)) {
            return 1.0; // Free shipping
        } else if ("platinum".equals(memberLevel)) {
            return 1.0; // Free express shipping
        }
        return 0.0;
    }
    
    public boolean hasExpressShipping(String memberLevel) {
        return "platinum".equals(memberLevel);
    }
}

class EventInvitationManager {
    public boolean canAttendSaleEvents(String memberLevel) {
        return !"".equals(memberLevel) && memberLevel != null;
    }
    
    public boolean canAttendVIPEvents(String memberLevel) {
        return "gold".equals(memberLevel) || "platinum".equals(memberLevel);
    }
    
    public boolean canAttendExclusiveEvents(String memberLevel) {
        return "platinum".equals(memberLevel);
    }
}

// Client usage example
class CheckoutProcess {
    public void processCheckout(double amount, String memberLevel) {
        DiscountCalculator discountCalc = new DiscountCalculator();
        RewardPointsCalculator pointsCalc = new RewardPointsCalculator();
        MemberBenefitDisplay benefitDisplay = new MemberBenefitDisplay();
        ShippingService shippingService = new ShippingService();
        
        double discount = discountCalc.calculateDiscount(amount, memberLevel);
        int points = pointsCalc.calculatePoints(amount, memberLevel);
        double shippingDiscount = shippingService.getShippingDiscount(memberLevel);
        
        System.out.println("Order amount: " + amount);
        System.out.println("Discount: " + discount);
        System.out.println("Final amount: " + (amount - discount));
        System.out.println("Points earned: " + points);
        System.out.println("Shipping discount: " + (shippingDiscount * 100) + "%");
        System.out.println(benefitDisplay.getBenefitDescription(memberLevel));
    }
}
